packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -837,6 +860,181 @@ function rerenderReducer<S, I, A>(
   return [newState, dispatch];
 }
 
+type MutableSourceState<Source, Snapshot> = {|
+  config: MutableSourceHookConfig<Source, Snapshot>,
+  destroy?: () => void,
+  snapshot: Snapshot,
+  source: MutableSource<any>,
+|};
+
+function useMutableSourceImpl<Source, Snapshot>(
+  hook: Hook,
+  source: MutableSource<Source>,
+  config: MutableSourceHookConfig<Source, Snapshot>,
+): Snapshot {
+  const {getSnapshot, subscribe} = config;
+  const version = source._getVersion();
+  const pendingExpirationTime = getPendingExpirationTime(
+    currentlyRenderingRoot,
+    source,
+  );
+
+  // Is it safe to read from this source during the current render?
+  let isSafeToReadFromSource = false;
+
+  if (pendingExpirationTime !== null) {
+    // If the source has pending updates, we can use the current render's expiration
+    // time to determine if it's safe to read again from the source.
+    const currentTime = requestCurrentTimeForUpdate();
+    const suspenseConfig = requestCurrentSuspenseConfig();
+    const expirationTime = computeExpirationForFiber(
+      currentTime,
+      currentlyRenderingFiber,
+      suspenseConfig,
+    );
+
+    isSafeToReadFromSource =
+      pendingExpirationTime === NoWork ||
+      pendingExpirationTime >= expirationTime;
+
+    if (!isSafeToReadFromSource) {
+      // Preserve the pending update if the current priority doesn't include it.
+      currentlyRenderingFiber.expirationTime = pendingExpirationTime;
+      markUnprocessedUpdateTime(pendingExpirationTime);
+    }
+  } else {
+    // If there are no pending updates, we should still check the work-in-progress version.
+    // It's possible this source (or the part we are reading from) has just not yet been subscribed to.
+    const lastReadVersion = getWorkInProgressVersion(source, isPrimaryRenderer);
+    if (lastReadVersion === null) {
+      // This is the only case where we need to actually update the version number.
+      // A changed version number for a new source will throw and restart the render,
+      // at which point the work-in-progress version Map will be reinitialized.
+      // Once a source has been subscribed to, we use expiration time to determine safety.
+      setWorkInProgressVersion(source, version, isPrimaryRenderer);
```

Comment:
> we only need to populate the work-in-progress version if this is a new mount, because updates don't rely on mutable state: they read from a queue

I don't think this reflects the case of a new source being used for an already-mounted component.

Reply:
Yeah sorry, by "mount" I mean either a new hook instance or whenever the source or config changes.